sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/math.scala

```diff
@@ -51,16 +54,28 @@ abstract class LeafMathExpression(c: Double, name: String)
  * @param f The math function.
  * @param name The short name of the function
  */
-abstract class UnaryMathExpression(f: Double => Double, name: String)
+abstract class AbstractUnaryMathExpression[T, U](name: String)
```

Comment:
Maybe we could don't let `BIN` inherit from `UnaryMathExpression`, they don't have much in common. How do yo think?